windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -0,0 +1,1367 @@
+(*
+  Name:             UpdateStateMachine
+  Copyright:        Copyright (C) SIL International.
+  Documentation:
+  Description:
+  Create Date:      2 Nov 2023
+
+  Modified Date:    2 Nov 2023
+  Authors:          rcruickshank
+  Related Files:
+  Dependencies:
+
+  Bugs:
+  Todo:
+  Notes: For the state diagram in mermaid ../BackgroundUpdateStateDiagram.md
+  History:
+*)
+unit Keyman.System.UpdateStateMachine;
+
+interface
+
+uses
+  System.Classes,
+  System.SysUtils,
+  System.UITypes,
+  System.IOUtils,
+  System.Types,
+  Vcl.Forms,
+  TypInfo,
+  KeymanPaths,
+  utilkmshell,
+
+  httpuploader,
+  Keyman.System.UpdateCheckResponse,
+  UfrmStartInstall,
+  UfrmStartInstallNow,
+  Keyman.System.ExecutionHistory,
+  UfrmDownloadProgress;
+
+const
+  CheckPeriod: Integer = 7; // Days between checking for updates
+
+type
+  EUpdateStateMachine = class(Exception);
+
+  TUpdateStateMachineResult = (oucUnknown, oucShutDown, oucSuccess, oucNoUpdates, oucUpdatesAvailable, oucFailure, oucOffline);
+
+  TUpdateState = (usIdle, usUpdateAvailable, usDownloading, usWaitingRestart, usInstalling, usRetry, usPostInstall);
+
+  { Keyboard Package Params }
+  TUpdateStateMachineParamsPackage = record
+    ID: string;
+    NewID: string;
+    Description: string;
+    OldVersion, NewVersion: string;
+    DownloadURL: string;
+    SavePath: string;
+    FileName: string;
+    DownloadSize: Integer;
+    Install: Boolean;
+  end;
+  { Main Keyman Program }
+  TUpdateStateMachineParamsKeyman = record
+    OldVersion, NewVersion: string;
+    DownloadURL: string;
+    SavePath: string;
+    FileName: string;
+    DownloadSize: Integer;
+    Install: Boolean;
+  end;
+
+  TUpdateStateMachineParams = record
+    Keyman: TUpdateStateMachineParamsKeyman;
+    Packages: array of TUpdateStateMachineParamsPackage;
+    Result: TUpdateStateMachineResult;
+  end;
+
+  TUpdateStateMachineDownloadParams = record
+    Owner: TfrmDownloadProgress;
+    TotalSize: Integer;
+    TotalDownloads: Integer;
+    StartPosition: Integer;
+  end;
+
+   // Forward declaration
+   TUpdateStateMachine = class;
+    { State Classes Update }
+
+   TStateClass = class of TState;
+
+  TState = class abstract
+  private
+    bucStateContext: TUpdateStateMachine;
+    procedure ChangeState(newState: TStateClass);
+
+  public
+    constructor Create(Context: TUpdateStateMachine);
+    procedure Enter; virtual; abstract;
+    procedure Exit; virtual; abstract;
+    procedure HandleCheck; virtual; abstract;
+    function  HandleKmShell : Integer; virtual; abstract;
```

Comment:
The base state does not need to be a pure interface. That is the methods do not need to be abstract. This would reduce all the no-op events in certain states. I like have it at development time as it helps me make sure I haven't missed a state in the matrix.